onnxruntime/core/providers/shared/node_unit.cc

```diff
@@ -0,0 +1,144 @@
+    return node_.OutputDefs();
+  }
+
+  const std::string& OpType() const noexcept override { return node_.OpType(); }
+  int SinceVersion() const noexcept override { return node_.SinceVersion(); }
+  const std::string& Domain() const noexcept override { return node_.Domain(); }
+  const Path& ModelPath() const noexcept override { return node_.ModelPath(); }
+  const std::string& Name() const noexcept override { return node_.Name(); }
+
+  const Node& GetNode() const noexcept override { return node_; }
+
+  // size_t GetInputEdgesCount() const noexcept override { return node_.GetInputEdgesCount(); }
+  NodeIndex Index() const noexcept override { return node_.Index(); }
+
+  ProviderType GetExecutionProviderType() const noexcept override { return node_.GetExecutionProviderType(); }
+
+  // Node::NodeConstIterator OutputNodesBegin() const noexcept override { return node_.OutputNodesBegin(); }
+
+  // Node::NodeConstIterator OutputNodesEnd() const noexcept override { return node_.OutputNodesEnd(); }
+
+  const std::vector<const Node*> GetAllNodes() const noexcept override { return all_nodes_; }
+
+  INodeUnit::Type UnitType() const noexcept override { return INodeUnit::Type::Node; }
+
+ private:
+  const Node& node_;
+  std::vector<const Node*> all_nodes_;
+};
+
+class QDQNodeUnit : public INodeUnit {
+ public:
+  QDQNodeUnit(const GraphViewer& graph_viewer, const QDQ::NodeGroup& qdq_group);
+
+  const ConstPointerContainer<std::vector<NodeArg*>> InputDefs() const noexcept override {
+    return ConstPointerContainer<std::vector<NodeArg*>>(input_defs_);
+  }
+
+  const ConstPointerContainer<std::vector<NodeArg*>> OutputDefs() const noexcept override {
+    return ConstPointerContainer<std::vector<NodeArg*>>(output_defs_);
+  }
+
+  const std::string& OpType() const noexcept override { return node_.OpType(); }
+  int SinceVersion() const noexcept override { return node_.SinceVersion(); }
+  const std::string& Domain() const noexcept override { return node_.Domain(); }
+  const Path& ModelPath() const noexcept override { return node_.ModelPath(); }
+  const std::string& Name() const noexcept override { return node_.Name(); }
+
+  const Node& GetNode() const noexcept override { return node_; }
+  NodeIndex Index() const noexcept override { return node_.Index(); }
+  ProviderType GetExecutionProviderType() const noexcept override { return node_.GetExecutionProviderType(); }
+
+  const std::vector<const Node*> GetAllNodes() const noexcept override { return all_nodes_; }
+
+  INodeUnit::Type UnitType() const noexcept override { return INodeUnit::Type::QDQ; }
+
+ private:
+  void init();
+  const GraphViewer& graph_viewer_;
+  const QDQ::NodeGroup qdq_group_;
+  const Node& node_;
+  std::vector<NodeArg*> input_defs_;
+  std::vector<NodeArg*> output_defs_;
+  std::vector<const Node*> all_nodes_;
+};
+
+// QUESTION/TODO, do we want to embed the graph_viewer into the QDQNodeGroup?
```

Comment:
just some thoughts.. is it a must to represent qdq_group with node indices here? 
If the reason is only because the qdq selector returns a selection of qdq node indices,  maybe we can add extra nodegroup struct represented by node* and convert qdq group(indices) from the selection result to qdq group(node*)? maybe in this way graph_viewer is no long needed here?

Reply:
You don't have to represent the node using index, pointer should also be fine, however, I would prefer a size_t instead of a pointer